File: 6KYU/Ease the StockBroker/solution.clj
(ns stockbroker.core
  (:require [clojure.string :as str]))

(defn is-double? [n]
  (some? (re-matches #"\d+(\.\d{1,3})" (str n))))

(defn is-integer? [n]
  (some? (re-matches #"(\d+)" (str n))))

(defn order-string->order-map [order]
  (let [[ticker quantity price status] (str/split order #" ")]
    {:ticker   ticker
     :quantity (read-string quantity)
     :price    (read-string price)
     :status   (if (some? status) status "")}))

(defn validate-order [order]
  (let [valid-quantity (is-integer? (:quantity order))
        valid-price (is-double? (:price order))
        valid-status (and (not= "" (:status order)) (str/includes? "BS" (:status order)))]
    (assoc order :valid (and valid-quantity valid-status valid-price))))

(defn invalid-order->string [order]
  (str (:ticker order) " "
       (:quantity order) " "
       (:price order)
       (if (= "" (:status order))
         (str " ")
         (str " " (:status order) " "))))

(defn balance [ords]
  (if (= "" ords)
    "Buy: 0 Sell: 0"
    (let [orders (->> (str/split ords #",")
                      (map str/trim)
                      (map order-string->order-map)
                      (map validate-order))
          valid-orders (filter #(true? (:valid %)) orders)
          invalid-orders (filter #(false? (:valid %)) orders)
          summary (reduce (fn [acc o]
                            (if (= "B" (:status o))
                              [(+ (first acc) (* (:quantity o) (:price o))) (second acc)]
                              [(first acc) (+ (second acc) (* (:quantity o) (:price o)))]))
                          [0 0] valid-orders)]
      (str "Buy: " (format "%.0f" (if (zero? (first summary))
                                    0.0
                                    (first summary)))
           " Sell: " (format "%.0f" (if (zero? (second summary))
                                      0.0
                                      (second summary)))
           (if (not= (count invalid-orders) 0)
             (str "; Badly formed " (count invalid-orders) ": "
                  (str/join ";" (map invalid-order->string invalid-orders))
                  ";"))))))

File: 6KYU/Ease the StockBroker/solution.go
package kata

import (
	"fmt"
	"strconv"
	"strings"
)

type Order struct {
	Stock  string
	Amount int
	Price  float64
	Buy    bool
}

func parseOrder(s string) (Order, string) {
	s = strings.Trim(s, " ")
	items := strings.Split(s, " ")
	if len(items) != 4 {
		return Order{}, s
	}

	stock := items[0]
	amount, amountErr := strconv.Atoi(items[1])
	price, priceErr := strconv.ParseFloat(items[2], 64)
	buy := items[3]

	if buy != "B" && buy != "S" {
		return Order{}, s
	}

	if !strings.ContainsRune(items[2], '.') || priceErr != nil {
		return Order{}, s
	}

	if amountErr != nil {
		return Order{}, s
	}

	return Order{stock, amount, price, buy == "B"}, ""
}

func BalanceStatement(input string) string {
	orders := strings.Split(input, ",")
	badlyFormedOrders := make([]string, 0)
	totalBuys := 0.0
	totalSells := 0.0

	for _, order := range orders {
		orderItems, err := parseOrder(order)

		if err != "" {
			badlyFormedOrders = append(badlyFormedOrders, err)
		} else {
			if orderItems.Buy {
				totalBuys += float64(orderItems.Amount) * orderItems.Price
			} else {
				totalSells += float64(orderItems.Amount) * orderItems.Price
			}
		}
	}

	summary := fmt.Sprintf("Buy: %0.f Sell: %0.f", totalBuys, totalSells)
	if len(badlyFormedOrders) > 0 {
		summary += fmt.Sprintf("; Badly formed %d: %s ;", len(badlyFormedOrders), strings.Join(badlyFormedOrders, " ;"))
	}

	return summary
}
